File: src/main/java/me/ag2s/epublib/util/IOUtil.java
package me.ag2s.epublib.util;

import java.io.ByteArrayOutputStream;
import java.io.Closeable;
import java.io.EOFException;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.Reader;
import java.io.StringWriter;
import java.io.Writer;
import java.net.HttpURLConnection;
import java.net.URLConnection;
import java.nio.ByteBuffer;
import java.nio.CharBuffer;
import java.nio.channels.ReadableByteChannel;
import java.nio.charset.Charset;
import me.ag2s.epublib.util.commons.io.IOConsumer;

public class IOUtil {
   private static final String TAG = IOUtil.class.getName();
   public static final int EOF = -1;
   public static final int DEFAULT_BUFFER_SIZE = 8192;
   private static final byte[] SKIP_BYTE_BUFFER = new byte[8192];
   private static char[] SKIP_CHAR_BUFFER;

   public static byte[] toByteArray(Reader in, String encoding) throws IOException {
      StringWriter out = new StringWriter();
      copy((Reader)in, (Writer)out);
      out.flush();
      return out.toString().getBytes(encoding);
   }

   public static byte[] toByteArray(InputStream in) throws IOException {
      ByteArrayOutputStream result = new ByteArrayOutputStream();
      copy((InputStream)in, (OutputStream)result);
      result.flush();
      return result.toByteArray();
   }

   public static byte[] toByteArray(InputStream in, int size) throws IOException {
      try {
         ByteArrayOutputStream result;
         if (size > 0) {
            result = new ByteArrayOutputStream(size);
         } else {
            result = new ByteArrayOutputStream();
         }

         copy((InputStream)in, (OutputStream)result);
         result.flush();
         return result.toByteArray();
      } catch (OutOfMemoryError var3) {
         return null;
      }
   }

   protected static int calcNewNrReadSize(int nrRead, int totalNrNread) {
      if (totalNrNread < 0) {
         return totalNrNread;
      } else {
         return totalNrNread > Integer.MAX_VALUE - nrRead ? -1 : totalNrNread + nrRead;
      }
   }

   public static void copy(InputStream in, OutputStream result) throws IOException {
      copy(in, result, 8192);
   }

   public static long copy(final InputStream input, final OutputStream output, final int bufferSize) throws IOException {
      return copyLarge(input, output, new byte[bufferSize]);
   }

   /** @deprecated */
   @Deprecated
   public static void copy(final InputStream input, final Writer output) throws IOException {
      copy(input, output, Charset.defaultCharset());
   }

   public static void copy(final InputStream input, final Writer output, final Charset inputCharset) throws IOException {
      InputStreamReader in = new InputStreamReader(input, inputCharset.name());
      copy((Reader)in, (Writer)output);
   }

   public static void copy(final InputStream input, final Writer output, final String inputCharsetName) throws IOException {
      copy(input, output, Charset.forName(inputCharsetName));
   }

   public static long copy(final Reader input, final Appendable output) throws IOException {
      return copy(input, output, CharBuffer.allocate(8192));
   }

   public static long copy(final Reader input, final Appendable output, final CharBuffer buffer) throws IOException {
      long count;
      int n;
      for(count = 0L; -1 != (n = input.read(buffer)); count += (long)n) {
         buffer.flip();
         output.append(buffer, 0, n);
      }

      return count;
   }

   /** @deprecated */
   @Deprecated
   public static void copy(final Reader input, final OutputStream output) throws IOException {
      copy(input, output, Charset.defaultCharset());
   }

   public static void copy(final Reader input, final OutputStream output, final Charset outputCharset) throws IOException {
      OutputStreamWriter out = new OutputStreamWriter(output, outputCharset.name());
      copy((Reader)input, (Writer)out);
      out.flush();
   }

   public static void copy(final Reader input, final OutputStream output, final String outputCharsetName) throws IOException {
      copy(input, output, Charset.forName(outputCharsetName));
   }

   public static int copy(final Reader input, final Writer output) throws IOException {
      long count = copyLarge(input, output);
      return count > 2147483647L ? -1 : (int)count;
   }

   public static long copyLarge(final InputStream input, final OutputStream output) throws IOException {
      return copy(input, output, 8192);
   }

   public static long copyLarge(final InputStream input, final OutputStream output, final byte[] buffer) throws IOException {
      long count = 0L;
      int n;
      if (input != null) {
         while(-1 != (n = input.read(buffer))) {
            output.write(buffer, 0, n);
            count += (long)n;
         }
      }

      return count;
   }

   public static long copyLarge(final InputStream input, final OutputStream output, final long inputOffset, final long length) throws IOException {
      return copyLarge(input, output, inputOffset, length, new byte[8192]);
   }

   public static long copyLarge(final InputStream input, final OutputStream output, final long inputOffset, final long length, final byte[] buffer) throws IOException {
      if (inputOffset > 0L) {
         skipFully(input, inputOffset);
      }

      if (length == 0L) {
         return 0L;
      } else {
         int bufferLength = buffer.length;
         int bytesToRead = bufferLength;
         if (length > 0L && length < (long)bufferLength) {
            bytesToRead = (int)length;
         }

         long totalRead = 0L;

         int read;
         while(bytesToRead > 0 && -1 != (read = input.read(buffer, 0, bytesToRead))) {
            output.write(buffer, 0, read);
            totalRead += (long)read;
            if (length > 0L) {
               bytesToRead = (int)Math.min(length - totalRead, (long)bufferLength);
            }
         }

         return totalRead;
      }
   }

   public static long copyLarge(final Reader input, final Writer output) throws IOException {
      return copyLarge(input, output, new char[8192]);
   }

   public static long copyLarge(final Reader input, final Writer output, final char[] buffer) throws IOException {
      long count;
      int n;
      for(count = 0L; -1 != (n = input.read(buffer)); count += (long)n) {
         output.write(buffer, 0, n);
      }

      return count;
   }

   public static long copyLarge(final Reader input, final Writer output, final long inputOffset, final long length) throws IOException {
      return copyLarge(input, output, inputOffset, length, new char[8192]);
   }

   public static long copyLarge(final Reader input, final Writer output, final long inputOffset, final long length, final char[] buffer) throws IOException {
      if (inputOffset > 0L) {
         skipFully(input, inputOffset);
      }

      if (length == 0L) {
         return 0L;
      } else {
         int bytesToRead = buffer.length;
         if (length > 0L && length < (long)buffer.length) {
            bytesToRead = (int)length;
         }

         long totalRead = 0L;

         int read;
         while(bytesToRead > 0 && -1 != (read = input.read(buffer, 0, bytesToRead))) {
            output.write(buffer, 0, read);
            totalRead += (long)read;
            if (length > 0L) {
               bytesToRead = (int)Math.min(length - totalRead, (long)buffer.length);
            }
         }

         return totalRead;
      }
   }

   public static long skip(final InputStream input, final long toSkip) throws IOException {
      if (toSkip < 0L) {
         throw new IllegalArgumentException("Skip count must be non-negative, actual: " + toSkip);
      } else {
         long remain;
         long n;
         for(remain = toSkip; remain > 0L; remain -= n) {
            n = (long)input.read(SKIP_BYTE_BUFFER, 0, (int)Math.min(remain, (long)SKIP_BYTE_BUFFER.length));
            if (n < 0L) {
               break;
            }
         }

         return toSkip - remain;
      }
   }

   public static long skip(final ReadableByteChannel input, final long toSkip) throws IOException {
      if (toSkip < 0L) {
         throw new IllegalArgumentException("Skip count must be non-negative, actual: " + toSkip);
      } else {
         ByteBuffer skipByteBuffer = ByteBuffer.allocate((int)Math.min(toSkip, (long)SKIP_BYTE_BUFFER.length));

         long remain;
         int n;
         for(remain = toSkip; remain > 0L; remain -= (long)n) {
            skipByteBuffer.position(0);
            skipByteBuffer.limit((int)Math.min(remain, (long)SKIP_BYTE_BUFFER.length));
            n = input.read(skipByteBuffer);
            if (n == -1) {
               break;
            }
         }

         return toSkip - remain;
      }
   }

   public static long skip(final Reader input, final long toSkip) throws IOException {
      if (toSkip < 0L) {
         throw new IllegalArgumentException("Skip count must be non-negative, actual: " + toSkip);
      } else {
         if (SKIP_CHAR_BUFFER == null) {
            SKIP_CHAR_BUFFER = new char[SKIP_BYTE_BUFFER.length];
         }

         long remain;
         long n;
         for(remain = toSkip; remain > 0L; remain -= n) {
            n = (long)input.read(SKIP_CHAR_BUFFER, 0, (int)Math.min(remain, (long)SKIP_BYTE_BUFFER.length));
            if (n < 0L) {
               break;
            }
         }

         return toSkip - remain;
      }
   }

   public static void skipFully(final InputStream input, final long toSkip) throws IOException {
      if (toSkip < 0L) {
         throw new IllegalArgumentException("Bytes to skip must not be negative: " + toSkip);
      } else {
         long skipped = skip(input, toSkip);
         if (skipped != toSkip) {
            throw new EOFException("Bytes to skip: " + toSkip + " actual: " + skipped);
         }
      }
   }

   public static void skipFully(final ReadableByteChannel input, final long toSkip) throws IOException {
      if (toSkip < 0L) {
         throw new IllegalArgumentException("Bytes to skip must not be negative: " + toSkip);
      } else {
         long skipped = skip(input, toSkip);
         if (skipped != toSkip) {
            throw new EOFException("Bytes to skip: " + toSkip + " actual: " + skipped);
         }
      }
   }

   public static void skipFully(final Reader input, final long toSkip) throws IOException {
      long skipped = skip(input, toSkip);
      if (skipped != toSkip) {
         throw new EOFException("Chars to skip: " + toSkip + " actual: " + skipped);
      }
   }

   public static int length(final byte[] array) {
      return array == null ? 0 : array.length;
   }

   public static int length(final char[] array) {
      return array == null ? 0 : array.length;
   }

   public static int length(final CharSequence csq) {
      return csq == null ? 0 : csq.length();
   }

   public static int length(final Object[] array) {
      return array == null ? 0 : array.length;
   }

   public static void close(final Closeable closeable) throws IOException {
      if (closeable != null) {
         closeable.close();
      }

   }

   public static void close(final Closeable... closeables) throws IOException {
      if (closeables != null) {
         for(Closeable closeable : closeables) {
            close(closeable);
         }
      }

   }

   public static void close(final Closeable closeable, final IOConsumer consumer) throws IOException {
      if (closeable != null) {
         try {
            closeable.close();
         } catch (IOException e) {
            if (consumer != null) {
               consumer.accept(e);
            }
         }
      }

   }

   public static void close(final URLConnection conn) {
      if (conn instanceof HttpURLConnection) {
         ((HttpURLConnection)conn).disconnect();
      }

   }

   public static String Stream2String(InputStream inputStream) {
      ByteArrayOutputStream result = new ByteArrayOutputStream();

      try {
         byte[] buffer = new byte[8192];

         int length;
         while((length = inputStream.read(buffer)) != -1) {
            result.write(buffer, 0, length);
         }

         return result.toString();
      } catch (Exception e) {
         return e.getLocalizedMessage();
      }
   }
}

File: src/main/kotlin/me/ag2s/epublib/util/IOUtil.kt
package me.ag2s.epublib.util

import java.io.Closeable
import java.io.ByteArrayOutputStream
import java.io.EOFException
import java.io.InputStream
import java.io.InputStreamReader
import java.io.OutputStream
import java.io.OutputStreamWriter
import java.io.Reader
import java.io.StringWriter
import java.io.Writer
import java.net.HttpURLConnection
import java.net.URLConnection
import java.nio.ByteBuffer
import java.nio.CharBuffer
import java.nio.channels.ReadableByteChannel
import java.nio.charset.Charset
import me.ag2s.epublib.util.commons.io.IOConsumer

object IOUtil {
    const val EOF = -1
    const val DEFAULT_BUFFER_SIZE = 8192
    private val skipByteBuffer = ByteArray(DEFAULT_BUFFER_SIZE)
    private var skipCharBuffer: CharArray? = null

    @JvmStatic fun toByteArray(input: Reader, encoding: String): ByteArray = StringWriter().use { output -> copy(input, output); output.flush(); output.toString().toByteArray(Charset.forName(encoding)) }
    @JvmStatic fun toByteArray(input: InputStream): ByteArray = ByteArrayOutputStream().use { output -> copy(input, output); output.flush(); output.toByteArray() }
    @JvmStatic fun toByteArray(input: InputStream, size: Int): ByteArray? = try { ByteArrayOutputStream(if (size > 0) size else 32).use { output -> copy(input, output); output.flush(); output.toByteArray() } } catch (_: OutOfMemoryError) { null }

    @JvmStatic fun calcNewNrReadSize(nrRead: Int, totalNrNread: Int): Int = if (totalNrNread < 0) totalNrNread else if (totalNrNread > Int.MAX_VALUE - nrRead) -1 else totalNrNread + nrRead
    @JvmStatic fun copy(input: InputStream, output: OutputStream) { copy(input, output, DEFAULT_BUFFER_SIZE) }
    @JvmStatic fun copy(input: InputStream, output: OutputStream, bufferSize: Int): Long = copyLarge(input, output, ByteArray(bufferSize))
    @JvmStatic fun copy(input: InputStream, output: Writer) { copy(input, output, Charset.defaultCharset()) }
    @JvmStatic fun copy(input: InputStream, output: Writer, inputCharset: Charset) { copy(InputStreamReader(input, inputCharset.name()), output) }
    @JvmStatic fun copy(input: InputStream, output: Writer, inputCharsetName: String) { copy(input, output, Charset.forName(inputCharsetName)) }
    @JvmStatic fun copy(input: Reader, output: Appendable): Long = copy(input, output, CharBuffer.allocate(DEFAULT_BUFFER_SIZE))
    @JvmStatic fun copy(input: Reader, output: Appendable, buffer: CharBuffer): Long { var count = 0L; var read: Int; while (input.read(buffer).also { read = it } != -1) { buffer.flip(); output.append(buffer, 0, read); count += read }; return count }
    @JvmStatic fun copy(input: Reader, output: OutputStream) { copy(input, output, Charset.defaultCharset()) }
    @JvmStatic fun copy(input: Reader, output: OutputStream, outputCharset: Charset) { val writer = OutputStreamWriter(output, outputCharset.name()); copy(input, writer); writer.flush() }
    @JvmStatic fun copy(input: Reader, output: OutputStream, outputCharsetName: String) { copy(input, output, Charset.forName(outputCharsetName)) }
    @JvmStatic fun copy(input: Reader, output: Writer): Int { val count = copyLarge(input, output); return if (count > Int.MAX_VALUE) -1 else count.toInt() }
    @JvmStatic fun copyLarge(input: InputStream, output: OutputStream): Long = copy(input, output, DEFAULT_BUFFER_SIZE)
    @JvmStatic fun copyLarge(input: InputStream?, output: OutputStream, buffer: ByteArray): Long { var count = 0L; if (input != null) { var read: Int; while (input.read(buffer).also { read = it } != -1) { output.write(buffer, 0, read); count += read } }; return count }
    @JvmStatic fun copyLarge(input: InputStream, output: OutputStream, offset: Long, length: Long): Long = copyLarge(input, output, offset, length, ByteArray(DEFAULT_BUFFER_SIZE))
    @JvmStatic fun copyLarge(input: InputStream, output: OutputStream, offset: Long, length: Long, buffer: ByteArray): Long { if (offset > 0) skipFully(input, offset); if (length == 0L) return 0; var bytesToRead = if (length in 1 until buffer.size.toLong()) length.toInt() else buffer.size; var total = 0L; while (bytesToRead > 0) { val read = input.read(buffer, 0, bytesToRead); if (read == -1) break; output.write(buffer, 0, read); total += read; if (length > 0) bytesToRead = minOf(length - total, buffer.size.toLong()).toInt() }; return total }
    @JvmStatic fun copyLarge(input: Reader, output: Writer): Long = copyLarge(input, output, CharArray(DEFAULT_BUFFER_SIZE))
    @JvmStatic fun copyLarge(input: Reader, output: Writer, buffer: CharArray): Long { var count = 0L; var read: Int; while (input.read(buffer).also { read = it } != -1) { output.write(buffer, 0, read); count += read }; return count }
    @JvmStatic fun copyLarge(input: Reader, output: Writer, offset: Long, length: Long): Long = copyLarge(input, output, offset, length, CharArray(DEFAULT_BUFFER_SIZE))
    @JvmStatic fun copyLarge(input: Reader, output: Writer, offset: Long, length: Long, buffer: CharArray): Long { if (offset > 0) skipFully(input, offset); if (length == 0L) return 0; var bytesToRead = if (length in 1 until buffer.size.toLong()) length.toInt() else buffer.size; var total = 0L; while (bytesToRead > 0) { val read = input.read(buffer, 0, bytesToRead); if (read == -1) break; output.write(buffer, 0, read); total += read; if (length > 0) bytesToRead = minOf(length - total, buffer.size.toLong()).toInt() }; return total }
    @JvmStatic fun skip(input: InputStream, toSkip: Long): Long { require(toSkip >= 0) { "Skip count must be non-negative, actual: $toSkip" }; var remaining = toSkip; while (remaining > 0) { val read = input.read(skipByteBuffer, 0, minOf(remaining, skipByteBuffer.size.toLong()).toInt()); if (read < 0) break; remaining -= read }; return toSkip - remaining }
    @JvmStatic fun skip(input: ReadableByteChannel, toSkip: Long): Long { require(toSkip >= 0) { "Skip count must be non-negative, actual: $toSkip" }; val buffer = ByteBuffer.allocate(minOf(toSkip, skipByteBuffer.size.toLong()).toInt()); var remaining = toSkip; while (remaining > 0) { buffer.position(0); buffer.limit(minOf(remaining, skipByteBuffer.size.toLong()).toInt()); val read = input.read(buffer); if (read == -1) break; remaining -= read }; return toSkip - remaining }
    @JvmStatic fun skip(input: Reader, toSkip: Long): Long { require(toSkip >= 0) { "Skip count must be non-negative, actual: $toSkip" }; val buffer = skipCharBuffer ?: CharArray(skipByteBuffer.size).also { skipCharBuffer = it }; var remaining = toSkip; while (remaining > 0) { val read = input.read(buffer, 0, minOf(remaining, buffer.size.toLong()).toInt()); if (read < 0) break; remaining -= read }; return toSkip - remaining }
    @JvmStatic fun skipFully(input: InputStream, toSkip: Long) { require(toSkip >= 0) { "Bytes to skip must not be negative: $toSkip" }; val skipped = skip(input, toSkip); if (skipped != toSkip) throw EOFException("Bytes to skip: $toSkip actual: $skipped") }
    @JvmStatic fun skipFully(input: ReadableByteChannel, toSkip: Long) { require(toSkip >= 0) { "Bytes to skip must not be negative: $toSkip" }; val skipped = skip(input, toSkip); if (skipped != toSkip) throw EOFException("Bytes to skip: $toSkip actual: $skipped") }
    @JvmStatic fun skipFully(input: Reader, toSkip: Long) { val skipped = skip(input, toSkip); if (skipped != toSkip) throw EOFException("Chars to skip: $toSkip actual: $skipped") }
    @JvmStatic fun length(array: ByteArray?): Int = array?.size ?: 0
    @JvmStatic fun length(array: CharArray?): Int = array?.size ?: 0
    @JvmStatic fun length(csq: CharSequence?): Int = csq?.length ?: 0
    @JvmStatic fun length(array: Array<out Any?>?): Int = array?.size ?: 0
    @JvmStatic fun close(closeable: Closeable?) { closeable?.close() }
    @JvmStatic fun close(vararg closeables: Closeable?) { closeables.forEach { close(it) } }
    @JvmStatic fun close(closeable: Closeable?, consumer: IOConsumer<java.io.IOException>?) { if (closeable != null) try { closeable.close() } catch (e: java.io.IOException) { consumer?.accept(e) } }
    @JvmStatic fun close(conn: URLConnection) { if (conn is HttpURLConnection) conn.disconnect() }
    @JvmStatic fun Stream2String(inputStream: InputStream): String = try { ByteArrayOutputStream().use { output -> val buffer = ByteArray(DEFAULT_BUFFER_SIZE); var length: Int; while (inputStream.read(buffer).also { length = it } != -1) output.write(buffer, 0, length); output.toString() } } catch (e: Exception) { e.localizedMessage }
}
